File: 0525-contiguous-array/0525-contiguous-array.js
var findMaxLength = function(nums) {
    nums.forEach((v, i) => {
        if (v == 0) nums[i] = -1;
    });

    let sumMap = new Map();
    let sum = 0;
    let maxLength = 0;

    sumMap.set(0, -1); // 누적 합이 0인 경우를 처리

    nums.forEach((num, i) => {
        sum += num;
        if (sumMap.has(sum)) {
            maxLength = Math.max(maxLength, i - sumMap.get(sum));
        } else {
            sumMap.set(sum, i);
        }
    });

    return maxLength;
};

File: 0525-contiguous-array/0525-contiguous-array.go
func findMaxLength(nums []int) int {
    sumMap := make(map[int]int)
    maxLength := 0;
    sum :=0;

    sumMap[0] = -1;

    for i, v := range nums {
     if v == 0 {
        v = -1
     }

        sum += v;

        if prevIndex, exists := sumMap[sum]; exists {
            maxLength = max(maxLength, i - prevIndex)
        } else {
            //없으면 map에 등록
            sumMap[sum] = i
        }

    }


    return maxLength;
}


func max(a,b int) int {
    if(a > b) {
        return a
    } else {
        return b
    }

}